### PR TITLE
Extract message emitters from Printer

### PR DESCRIPTION
**Pull request by MichaReiser**
This PR extracts the emitters for serializing `Messages` from the `Printer` in the CLI to `ruff`. The reason for moving the implementations to ruff is so that I use the text emitter to write the linter snapshots instead of serializing the `Diagnostic` to YAML. Using the rendered text output has the advantage that we can do refactors like replacing `Location` with `TextSize` without affecting the snapshot outputs, which gives us confidence in the refactor. 

I used this chance to:

* Add tests for all emitters
* Rewrite some emitters to use `write` directly rather than using `format` and then writing the formatted string. 

## Considerations

* Introducing a shared trait for the emitters isn't necessary from an implementation point of view because the `Printer` calls `emit` statically on all emitters (it isn't using dynamic dispatching). However, it helped to share logic between tests and it ensures that all emitters are structured similarly.  
* Implementing all emitters in the `ruff` crate isn't strictly necessary. For example, I don't plan to use the `junit` emitter for the test infrastructure. I still decided to put them all into the same module for better discoverability. 

## Performance

Avoiding `format` and `to_string` improves performance a bit

<pre><span style="color:#005FD7">hyperfine</span> <span style="color:#00AFFF">--warmup</span> <span style="color:#00AFFF">10</span> \
          <span style="color:#999900">&quot;./target/release/ruff ./crates/ruff/resources/test/cpython/ -e --select=ALL&quot;</span> \
          <span style="color:#999900">&quot;./ruff-main ./crates/ruff/resources/test/cpython/ -e --select=ALL&quot;</span>
<b>Benchmark 1</b>: ./target/release/ruff ./crates/ruff/resources/test/cpython/ -e --select=ALL
  Time (<span style="color:#26A269"><b>mean</b></span> ± <span style="color:#26A269">σ</span>):     <span style="color:#26A269"><b>442.6 ms</b></span> ± <span style="color:#26A269"> 10.0 ms</span>    [User: <span style="color:#12488B">722.9 ms</span>, System: <span style="color:#12488B">509.2 ms</span>]
  Range (<span style="color:#2AA1B3">min</span> … <span style="color:#A347BA">max</span>):   <span style="color:#2AA1B3">422.4 ms</span> … <span style="color:#A347BA">456.9 ms</span>    10 runs
 
<b>Benchmark 2</b>: ./ruff-main ./crates/ruff/resources/test/cpython/ -e --select=ALL
  Time (<span style="color:#26A269"><b>mean</b></span> ± <span style="color:#26A269">σ</span>):     <span style="color:#26A269"><b>456.9 ms</b></span> ± <span style="color:#26A269">  7.0 ms</span>    [User: <span style="color:#12488B">724.0 ms</span>, System: <span style="color:#12488B">507.0 ms</span>]
  Range (<span style="color:#2AA1B3">min</span> … <span style="color:#A347BA">max</span>):   <span style="color:#2AA1B3">448.1 ms</span> … <span style="color:#A347BA">469.2 ms</span>    10 runs
 
<b>Summary</b>
  &apos;<span style="color:#2AA1B3">./target/release/ruff ./crates/ruff/resources/test/cpython/ -e --select=ALL</span>&apos; ran
<span style="color:#26A269"><b>    1.03</b></span> ± <span style="color:#26A269">0.03</span> times faster than &apos;<span style="color:#A347BA">./ruff-main ./crates/ruff/resources/test/cpython/ -e --select=ALL</span>&apos;
</pre>

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3895** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/3895" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/charliermarsh/ruff/3895?utm_source=stack-comment).

File: crates/ruff/src/message/mod.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+use serde::{Deserialize, Serialize};
+pub use text::TextEmitter;
+
+use crate::jupyter::JupyterIndex;
+use ruff_diagnostics::{Diagnostic, DiagnosticKind, Fix};
+use ruff_python_ast::source_code::Locator;
+use ruff_python_ast::types::Range;
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Message {
```
Comment (MichaReiser) [inline crates/ruff/src/message/mod.rs:32]: Copied from `src/message.rs`

File: crates/ruff/src/message/mod.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    for message in messages {
+        grouped_messages
+            .entry(message.filename.as_str())
+            .or_insert_with(Vec::new)
+            .push(message);
+    }
+    grouped_messages
+}
+
+pub trait Emitter {
```
Comment (MichaReiser) [inline crates/ruff/src/message/mod.rs:126]: This is new

File: crates/ruff/src/message/github.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+use crate::message::{Emitter, EmitterContext, Message};
+use crate::registry::AsRule;
+use std::io::Write;
+
+/// Generate error workflow command in GitHub Actions format.
+/// See: [GitHub documentation](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message)
+#[derive(Default)]
+pub struct GithubEmitter;
+
+impl Emitter for GithubEmitter {
```
Comment (charliermarsh) [inline crates/ruff/src/message/github.rs:11]: My read on the Rust naming rules is that this is correct (over `GitHubEmitter`).

**Review by charliermarsh — APPROVED** [review]

File: crates/ruff/src/message/github.rs (hunk 3 above)
Comment (MichaReiser) [reply crates/ruff/src/message/github.rs:11]: I don't know. I choice this naming to keep it consistent with `SerializationFormat::Github`

**Pull request merged**